src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/CSharpOrVisualBasicFileRenameHandler.cs

```diff
@@ -0,0 +1,243 @@
+            {
+                return;
+            }
+
+            Project project = GetCurrentProject();
+            if (project is null)
+                return;
+
+            // see if the current project Roslyn compilations and if the language in that compilation is case-sensitive
+            (bool success, bool isCaseSensitive) = await TryDetermineIfCompilationIsCaseSensitiveAsync(project);
+            if (!success)
+                return;
+
+            // Check that the new name is a valid identifier in the current programming language
+            string oldName = Path.GetFileNameWithoutExtension(oldFilePath);
+            string newName = Path.GetFileNameWithoutExtension(newFilePath);
+            if (!CanHandleRename(oldName, newName, isCaseSensitive))
+                return;
+
+            // Check if there are any symbols that need to be renamed
+            (success, _) = await TryGetSymbolToRename(oldName, newFilePath, isCaseSensitive, project, default);
+            if (!success)
+                return;
+
+            // Ask if the user wants to rename the symbol
+            bool userConfirmed = await CheckUserConfirmation(oldName);
+            if (!userConfirmed)
+                return;
+
+            // Try and apply the changes to the current solution
+            await _projectVsServices.ThreadingService.SwitchToUIThread();
+
+            bool renamedSolutionApplied = _waitService.WaitForAsyncOperation(
+                title: VSResources.Performing_Rename,
+                message: string.Format(CultureInfo.CurrentCulture, VSResources.Renaming_type_from_0_to_1, oldName, newName),
+                allowCancel: true,
+                token =>
+                    // Do not let the project close until this completes
+                    _unconfiguredProjectTasksService.LoadedProjectAsync(
+                    async () =>
+                    {
+                        // Perform the rename operation
+                        Solution renamedSolution = await GetRenamedSolutionAsync(oldName, newFilePath, isCaseSensitive, GetCurrentProject(), token);
+                        if (renamedSolution == null)
+                            return false;
+
+                        // Try and apply the changes to the current solution
+                        return _roslynServices.ApplyChangesToSolution(renamedSolution.Workspace, renamedSolution);
+                    }));
+
+
+            // Notify the user if the rename could not be performed
+            if (!renamedSolutionApplied)
+            {
+                string failureMessage = string.Format(CultureInfo.CurrentCulture, VSResources.RenameSymbolFailed, oldName);
+                _userNotificationServices.ShowWarning(failureMessage);
+            }
+
+            Project GetCurrentProject()
+            {
+                foreach (Project proj in _workspace.CurrentSolution.Projects)
```

Comment:
What happens in multi-targeting projects? We just randomly choose one? Perhaps we should rename in both?

Reply:
Roslyn doesn't have the concept of multi-targeting in the workspace and its apis already understand shared projects and linked files so its safe to pick the first thing we find

Reply:
Can you expand on the "safe" to pick to the first thing we find? 

I have two projects; net45, net46. Someone references my net45 version and consumes the type in the file I just renamed. If we pick net46, we won't rename this usage. If its consumed differently by different projects, then we will only rename one side of these usages.

This might be a tough problem to solve, and maybe we pick a consistent behavior (always rename the first TFM, but not others) but relying on the order of what CurrentSolution.Projects returns does not seem like a safe approach.

Reply:
> but relying on the order of what CurrentSolution.Projects returns does not seem like a safe approach.

I will attempt to argue that this is safe because the roslyn workspace already needs to handle this case for linked files today.

Talked to @jasonmalinowski 	to confirm this behavior.

Multi-TFM projects are treated the same linked-files by the roslyn workspace. A linked file can be contained in `N` number of projects but if I grab a linked file from an arbitrary project and do a rename Roslyn understands that operation from the perspective of whatever the IDE's 'context' is.

Say I have a multi-TFM project like this:
```XML
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFrameworks>net472;netstandard2.0</TargetFrameworks>
  </PropertyGroup>
</Project>
```

This will be represented as two projects by the roslyn workspace, but either Document that I get, when I go to examine it, will appear in whatever context the editor is currently in.  So if I grab the document from ConsoleApp17_net472 but my IDE looks like this:
![image](https://user-images.githubusercontent.com/9797472/53991247-05c21180-40df-11e9-8dba-82a50ef81cf7.png)

The semantic information that I get will be as if it was from ConsoleApp17_netstandard2.0.

Admittedly, since we can only be in one 'context' at a time in the IDE there are cases where a rename with one 'context' active will mess up a different 'context' (as has been true for the last 5 years in linked files and conditional compilation symbols).

So we can

1. Block this PR until roslyn provides and API that allows us to reason about all TFMs simultaneously in a rename
2. Do not offer to rename the type if we detect that we are in a linked-files/multi-TFM situation (detect the same file path in multiple project)
3. Allow this behavior as is since legacy has had this same behavior with linked files for the last 3 years.

Reply:
I'm okay with the behavior, just document that in the code - its not clear that's what it does. 

It's entirely unobvious that (at least to me) that if I grab a symbol from a document that it's going to return a symbol from a completely different project.

Reply:
👍